// @aries-framework/indy-sdk/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndySdkModuleConfig = exports.IndySdkModule = exports.IndySdkVerifierService = exports.IndySdkIssuerService = exports.IndySdkHolderService = exports.IndySdkAnonCredsRegistry = exports.IndySdkStorageService = exports.IndySdkWallet = exports.IndySdkSovDidResolver = exports.IndySdkSovDidRegistrar = void 0;
// Dids
var dids_1 = require("./dids");
Object.defineProperty(exports, "IndySdkSovDidRegistrar", { enumerable: true, get: function () { return dids_1.IndySdkSovDidRegistrar; } });
Object.defineProperty(exports, "IndySdkSovDidResolver", { enumerable: true, get: function () { return dids_1.IndySdkSovDidResolver; } });
// Wallet
var wallet_1 = require("./wallet");
Object.defineProperty(exports, "IndySdkWallet", { enumerable: true, get: function () { return wallet_1.IndySdkWallet; } });
// Storage
var storage_1 = require("./storage");
Object.defineProperty(exports, "IndySdkStorageService", { enumerable: true, get: function () { return storage_1.IndySdkStorageService; } });
// AnonCreds
var anoncreds_1 = require("./anoncreds");
Object.defineProperty(exports, "IndySdkAnonCredsRegistry", { enumerable: true, get: function () { return anoncreds_1.IndySdkAnonCredsRegistry; } });
Object.defineProperty(exports, "IndySdkHolderService", { enumerable: true, get: function () { return anoncreds_1.IndySdkHolderService; } });
Object.defineProperty(exports, "IndySdkIssuerService", { enumerable: true, get: function () { return anoncreds_1.IndySdkIssuerService; } });
Object.defineProperty(exports, "IndySdkVerifierService", { enumerable: true, get: function () { return anoncreds_1.IndySdkVerifierService; } });
// Module
var IndySdkModule_1 = require("./IndySdkModule");
Object.defineProperty(exports, "IndySdkModule", { enumerable: true, get: function () { return IndySdkModule_1.IndySdkModule; } });
var IndySdkModuleConfig_1 = require("./IndySdkModuleConfig");
Object.defineProperty(exports, "IndySdkModuleConfig", { enumerable: true, get: function () { return IndySdkModuleConfig_1.IndySdkModuleConfig; } });
