// @aries-framework/core/build/modules/proofs/protocol/v1/handlers/V1RequestPresentationHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.V1RequestPresentationHandler = void 0;
const models_1 = require("../../../../../agent/models");
const ServiceDecorator_1 = require("../../../../../decorators/service/ServiceDecorator");
const storage_1 = require("../../../../../storage");
const routing_1 = require("../../../../routing");
const messages_1 = require("../messages");
class V1RequestPresentationHandler {
    constructor(proofProtocol) {
        this.supportedMessages = [messages_1.V1RequestPresentationMessage];
        this.proofProtocol = proofProtocol;
    }
    async handle(messageContext) {
        const proofRecord = await this.proofProtocol.processRequest(messageContext);
        const shouldAutoRespond = await this.proofProtocol.shouldAutoRespondToRequest(messageContext.agentContext, {
            proofRecord,
            requestMessage: messageContext.message,
        });
        if (shouldAutoRespond) {
            return await this.acceptRequest(proofRecord, messageContext);
        }
    }
    async acceptRequest(proofRecord, messageContext) {
        messageContext.agentContext.config.logger.info(`Automatically sending presentation with autoAccept on`);
        if (messageContext.connection) {
            const { message } = await this.proofProtocol.acceptRequest(messageContext.agentContext, {
                proofRecord,
            });
            return new models_1.OutboundMessageContext(message, {
                agentContext: messageContext.agentContext,
                connection: messageContext.connection,
                associatedRecord: proofRecord,
            });
        }
        else if (messageContext.message.service) {
            const { message } = await this.proofProtocol.acceptRequest(messageContext.agentContext, {
                proofRecord,
            });
            const routingService = messageContext.agentContext.dependencyManager.resolve(routing_1.RoutingService);
            const routing = await routingService.getRouting(messageContext.agentContext);
            const ourService = new ServiceDecorator_1.ServiceDecorator({
                serviceEndpoint: routing.endpoints[0],
                recipientKeys: [routing.recipientKey.publicKeyBase58],
                routingKeys: routing.routingKeys.map((key) => key.publicKeyBase58),
            });
            const recipientService = messageContext.message.service;
            // Set and save ~service decorator to record (to remember our verkey)
            message.service = ourService;
            const didCommMessageRepository = messageContext.agentContext.dependencyManager.resolve(storage_1.DidCommMessageRepository);
            await didCommMessageRepository.saveOrUpdateAgentMessage(messageContext.agentContext, {
                agentMessage: message,
                associatedRecordId: proofRecord.id,
                role: storage_1.DidCommMessageRole.Sender,
            });
            return new models_1.OutboundMessageContext(message, {
                agentContext: messageContext.agentContext,
                serviceParams: {
                    service: recipientService.resolvedDidCommService,
                    senderKey: message.service.resolvedDidCommService.recipientKeys[0],
                },
            });
        }
        messageContext.agentContext.config.logger.error(`Could not automatically create presentation`);
    }
}
exports.V1RequestPresentationHandler = V1RequestPresentationHandler;
